pandas/core/arrays/string_arrow.py

```diff
@@ -685,6 +691,19 @@ def value_counts(self, dropna: bool = True) -> Series:
 
         return Series(counts, index=index).astype("Int64")
 
+    def astype(self, dtype, copy=True):
+        dtype = pandas_dtype(dtype)
+
+        if is_dtype_equal(dtype, self.dtype):
+            if copy:
+                return self.copy()
+            return self
+
+        elif hasattr(dtype, "__from_arrow__"):
+            return dtype.__from_arrow__(self._data)
```

Comment:
I don't think we should use this method like this: `__from_arrow__` is meant to convert data from arrow for that specific dtype, and doesn't necessarily need to include any casting logic (I think)

Reply:
Alternative logic could be something like:

```
elif isinstance(dtype, NumericDtype):
    data = self._data.cast(pa.from_numpy_dtype(dtype.numpy_dtype))
    return dtype.__from_arrow__(data)
```

that would specifically work for the numeric masked arrays, and not rely on ``__from_arrow__`` to do the casting (but rely on pyarrow for that). 

This would already support casting to nullable integer / float to get the tests passing.

An example why we can rely in general on `__from_arrow__` is eg if we would do here `string_array.astype(pd.PeriodDtype("D"))`, which would fail in ``__from_arrow__

Reply:
> An example why we can rely in general on `__from_arrow__` is eg if we would do here `string_array.astype(pd.PeriodDtype("D"))`, which would fail in ``**from_arrow**

indeed, we have a similar with StringArray as well #40566 where 

```
        elif isinstance(dtype, ExtensionDtype):
            cls = dtype.construct_array_type()
            return cls._from_sequence(self, dtype=dtype, copy=copy)
```
in many cases fails as \_from_sequence, like \_\_from_arrow__, does not support casting in many cases.